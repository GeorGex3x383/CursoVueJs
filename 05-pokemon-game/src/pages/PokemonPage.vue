<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1 class="fade-in">¿Quién es este pokémon?</h1>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />
        <template v-if="showAnswer" class="fade-in">
            <h2>{{message}}</h2>
            <button @click="newGame">Nuevo Juego</button>
        </template>
    </div>
</template>

<script>

import PokemonPicture from '../components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data(){
        return{
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixedPokemonArray(){
            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[rndInt]
        },
        checkAnswer(selectedId){
            this.showPokemon = true
            this.showAnswer = true
            if(selectedId === this.pokemon.id){
                this.message = `Correcto, el pokemón era ${this.pokemon.name }`
            } else {
                this.message = `Incorrecto, el pokemón era ${this.pokemon.name }`
            }
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixedPokemonArray()
        }
    },
    mounted(){
        this.mixedPokemonArray()
    }
}
</script>
