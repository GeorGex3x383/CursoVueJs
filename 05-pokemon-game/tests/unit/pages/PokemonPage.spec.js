import {shallowMount, mount} from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import {pokemons} from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {

    let wrapper;

    beforeEach( () => {
        wrapper = shallowMount(PokemonPage)
    })

    test('debe de llamar mixPokemonArr al montar', () => {

        const mixedPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixedPokemonArray')
        shallowMount(PokemonPage)
        expect(mixedPokemonArraySpy).toHaveBeenCalled()

    })

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {

        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

    })

    test('debe de mostrar los componentes de PokemonPicture y PokemonOption', () => {

        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()

        expect(picture.attributes('pokemonid')).toBe('1')
        expect(options.attributes('pokemons')).toBeTruthy()

    })

    test('pruebas con checkAnswer', async() => {

        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, el pokemón era ${pokemons[0].name}`)
        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.message).toBe(`Incorrecto, el pokemón era ${pokemons[0].name }`)

    })

})