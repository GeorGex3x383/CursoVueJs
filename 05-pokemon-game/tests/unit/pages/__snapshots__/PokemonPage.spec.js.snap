// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1 class="fade-in">¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
